File: src/wwwroot/js/drawflowinterop.js
export class DrawflowInterop {
    constructor() {
        this.instances = {};
    }

    create(elementId, options) {
        // Use direct id selector for interop
        const selector = `#${elementId}`;
        const container = document.querySelector(selector);

        const editor = new Drawflow(container);

        if (options) {
            // If options is a string (from Blazor), parse it
            if (typeof options === 'string') {
                try {
                    options = JSON.parse(options);
                } catch (e) {
                    console.error('Failed to parse Drawflow options:', e);
                }
            }
            Object.assign(editor, options);
        }

        editor.start();
        this.instances[elementId] = editor;
    }

    addNode(elementId, name, inputs, outputs, posX, posY, className, data, html) {
        const editor = this.instances[elementId];
        return editor.addNode(name, inputs, outputs, posX, posY, className, data, html);
    }

    removeNode(elementId, id) {
        const editor = this.instances[elementId];
        editor.removeNodeId('node-' + id); // Always a string
    }

    addConnection(elementId, outId, inId, outClass, inClass) {
        const editor = this.instances[elementId];
        editor.addConnection(outId, inId, outClass, inClass);
    }

    export(elementId) {
        const editor = this.instances[elementId];
        const data = editor.export();
        return JSON.stringify(data);
    }

    exportAsJson(elementId) {
        const editor = this.instances[elementId];
        const data = editor.export();
        return JSON.stringify(data);
    }

    import(elementId, json) {
        const editor = this.instances[elementId];
        editor.import(JSON.parse(json));
    }

    destroy(elementId) {
        const editor = this.instances[elementId];
        if (editor) {
            editor.destroy();
            delete this.instances[elementId];
        }
    }

    createObserver(elementId) {
        // This method is used by the EventListeningInterop base class
        // to observe DOM changes for the element
        // Use direct id selector for observer
        const selector = `#${elementId}`;
        const element = document.querySelector(selector);
        if (element) {
            // Create a MutationObserver to watch for changes
            const observer = new MutationObserver((mutations) => {
                mutations.forEach((mutation) => {
                    if (mutation.type === 'childList') {
                        // Handle DOM changes if needed
                    }
                });
            });
            
            observer.observe(element, {
                childList: true,
                subtree: true
            });
        }
    }

    addEventListener(elementId, eventName, dotNetCallback) {
        const editor = this.instances[elementId];
        editor.on(eventName, (...args) => {
            const json = JSON.stringify(args);
            dotNetCallback.invokeMethodAsync('Invoke', json);
        });
    }

    zoomIn(elementId) {
        const editor = this.instances[elementId];
        editor.zoom_in();
    }

    zoomOut(elementId) {
        const editor = this.instances[elementId];
        editor.zoom_out();
    }

    addModule(elementId, name) {
        const editor = this.instances[elementId];
        editor.addModule(name);
    }

    changeModule(elementId, name) {
        const editor = this.instances[elementId];
        editor.changeModule(name);
    }

    removeModule(elementId, name) {
        const editor = this.instances[elementId];
        editor.removeModule(name);
    }

    getNodeFromId(elementId, id) {
        const editor = this.instances[elementId];
        return editor.getNodeFromId(id);
    }

    getNodesFromName(elementId, name) {
        const editor = this.instances[elementId];
        return editor.getNodesFromName(name);
    }

    updateNodeData(elementId, id, data) {
        const editor = this.instances[elementId];
        editor.updateNodeDataFromId(id, data);
    }

    addNodeInput(elementId, id) {
        const editor = this.instances[elementId];
        editor.addNodeInput(id); // Always a string
    }

    addNodeOutput(elementId, id) {
        const editor = this.instances[elementId];
        editor.addNodeOutput(id); // Always a string
    }

    removeNodeInput(elementId, id, inputClass) {
        const editor = this.instances[elementId];
        editor.removeNodeInput(id, inputClass); // Always a string
    }

    removeNodeOutput(elementId, id, outputClass) {
        const editor = this.instances[elementId];
        editor.removeNodeOutput(id, outputClass); // Always a string
    }

    removeSingleConnection(elementId, outId, inId, outClass, inClass) {
        const editor = this.instances[elementId];
        editor.removeSingleConnection(outId, inId, outClass, inClass);
    }

    updateConnectionNodes(elementId, id) {
        const editor = this.instances[elementId];
        editor.updateConnectionNodes('node-' + id); // Always a string
    }

    removeConnectionNodeId(elementId, id) {
        const editor = this.instances[elementId];
        editor.removeConnectionNodeId('node-' + id); // Always a string
    }

    getModuleFromNodeId(elementId, id) {
        const editor = this.instances[elementId];
        return editor.getModuleFromNodeId(id);
    }

    clearModuleSelected(elementId) {
        const editor = this.instances[elementId];
        editor.clearModuleSelected();
    }

    clear(elementId) {
        const editor = this.instances[elementId];
        editor.clear();
    }

    // Additional methods from Drawflow documentation
    setZoom(elementId, zoom) {
        const editor = this.instances[elementId];
        editor.zoom = zoom;
    }

    getZoom(elementId) {
        const editor = this.instances[elementId];
        return editor.zoom;
    }

    centerNode(elementId, id) {
        const editor = this.instances[elementId];
        editor.centerNode(id);
    }

    getNodePosition(elementId, id) {
        const editor = this.instances[elementId];
        const node = editor.getNodeFromId(id);
        if (node) {
            return { x: node.pos_x, y: node.pos_y };
        }
        return null;
    }

    setNodePosition(elementId, id, posX, posY) {
        const editor = this.instances[elementId];
        const node = editor.getNodeFromId(id);
        if (node) {
            node.pos_x = posX;
            node.pos_y = posY;
            editor.updateConnectionNodes(id);
        }
    }

    getNodeHtml(elementId, id) {
        const editor = this.instances[elementId];
        const node = editor.getNodeFromId(id);
        return node ? node.html : '';
    }

    setNodeHtml(elementId, id, html) {
        const editor = this.instances[elementId];
        const node = editor.getNodeFromId(id);
        if (node) {
            node.html = html;
            editor.updateConnectionNodes(id);
        }
    }

    getNodeClass(elementId, id) {
        const editor = this.instances[elementId];
        const node = editor.getNodeFromId(id);
        return node ? node.class : '';
    }

    setNodeClass(elementId, id, className) {
        const editor = this.instances[elementId];
        const node = editor.getNodeFromId(id);
        if (node) {
            node.class = className;
            editor.updateConnectionNodes(id);
        }
    }

    getNodeName(elementId, id) {
        const editor = this.instances[elementId];
        const node = editor.getNodeFromId(id);
        return node ? node.name : '';
    }

    setNodeName(elementId, id, name) {
        const editor = this.instances[elementId];
        const node = editor.getNodeFromId(id);
        if (node) {
            node.name = name;
            editor.updateConnectionNodes(id);
        }
    }

    getNodeConnections(elementId, id) {
        const editor = this.instances[elementId];
        const node = editor.getNodeFromId(id);
        if (node) {
            const connections = [];
            if (node.inputs) {
                Object.keys(node.inputs).forEach(inputKey => {
                    if (node.inputs[inputKey].connections) {
                        connections.push(...node.inputs[inputKey].connections);
                    }
                });
            }
            if (node.outputs) {
                Object.keys(node.outputs).forEach(outputKey => {
                    if (node.outputs[outputKey].connections) {
                        connections.push(...node.outputs[outputKey].connections);
                    }
                });
            }
            return connections;
        }
        return [];
    }

    isNodeSelected(elementId, id) {
        const editor = this.instances[elementId];
        return editor.selectedNodes && editor.selectedNodes.includes(id);
    }

    selectNode(elementId, id) {
        const editor = this.instances[elementId];
        if (!editor.selectedNodes) {
            editor.selectedNodes = [];
        }
        if (!editor.selectedNodes.includes(id)) {
            editor.selectedNodes.push(id);
        }
    }

    unselectNode(elementId, id) {
        const editor = this.instances[elementId];
        if (editor.selectedNodes) {
            const index = editor.selectedNodes.indexOf(id);
            if (index > -1) {
                editor.selectedNodes.splice(index, 1);
            }
        }
    }

    getSelectedNodes(elementId) {
        const editor = this.instances[elementId];
        return editor.selectedNodes || [];
    }

    clearSelectedNodes(elementId) {
        const editor = this.instances[elementId];
        editor.selectedNodes = [];
    }

    getCurrentModule(elementId) {
        const editor = this.instances[elementId];
        return editor.module || 'Home';
    }

    getModules(elementId) {
        const editor = this.instances[elementId];
        return Object.keys(editor.drawflow || {});
    }

    isEditMode(elementId) {
        const editor = this.instances[elementId];
        return editor.editMode !== false;
    }

    setEditMode(elementId, editMode) {
        const editor = this.instances[elementId];
        editor.editMode = editMode;
    }

    // Batch operation example (optional)
    removeNodes(elementId, ids) {
        const editor = this.instances[elementId];
        ids.forEach(id => editor.removeNodeId('node-' + id));
    }
}

window.DrawflowInterop = new DrawflowInterop();

File: src/Soenneker.Blazor.Drawflow/DrawflowInterop.cs
using Microsoft.AspNetCore.Components;
using Microsoft.JSInterop;
using Soenneker.Asyncs.Initializers;
using Soenneker.Blazor.Drawflow.Abstract;
using Soenneker.Blazor.Drawflow.Dtos;
using Soenneker.Blazor.Drawflow.Options;
using Soenneker.Blazor.Drawflow.Utils;
using Soenneker.Blazor.Utils.ModuleImport.Abstract;
using Soenneker.Blazor.Utils.ResourceLoader.Abstract;
using Soenneker.Extensions.CancellationTokens;
using Soenneker.Extensions.Enumerable;
using Soenneker.Utils.CancellationScopes;
using Soenneker.Utils.Json;
using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;

namespace Soenneker.Blazor.Drawflow;

/// <inheritdoc cref="IDrawflowInterop"/>
public sealed class DrawflowInterop : IDrawflowInterop
{
    private const string _modulePath = "/_content/Soenneker.Blazor.Drawflow/js/drawflowinterop.js";

    private readonly IResourceLoader _resourceLoader;
    private readonly IModuleImportUtil _moduleImportUtil;

    private readonly AsyncInitializer _interopInitializer;
    private readonly AsyncInitializer<bool> _styleInitializer;
    private readonly AsyncInitializer<bool> _scriptInitializer;
    private const string _drawflowVariable = "Drawflow";

    private readonly CancellationScope _cancellationScope = new();

    public DrawflowInterop(IResourceLoader resourceLoader, IModuleImportUtil moduleImportUtil)
    {
        _resourceLoader = resourceLoader;
        _moduleImportUtil = moduleImportUtil;
        _styleInitializer = new AsyncInitializer<bool>(InitializeStyle);
        _scriptInitializer = new AsyncInitializer<bool>(InitializeScript);
        _interopInitializer = new AsyncInitializer(InitializeInterop);
    }

    private static string NormalizeContentUri(string uri)
    {
        if (string.IsNullOrEmpty(uri) || uri.Contains("://", StringComparison.Ordinal))
            return uri;

        return uri[0] == '/' ? uri : "/" + uri;
    }

    private async ValueTask InitializeStyle(bool useCdn, CancellationToken token)
    {
        (string uri, string? integrity) style = DrawflowUtil.GetUriAndIntegrityForStyle(useCdn);
        await _resourceLoader.LoadStyle(NormalizeContentUri(style.uri), style.integrity, cancellationToken: token);
    }

    private async ValueTask InitializeScript(bool useCdn, CancellationToken token)
    {
        (string uri, string? integrity) script = DrawflowUtil.GetUriAndIntegrityForScript(useCdn);
        await _resourceLoader.LoadScriptAndWaitForVariable(NormalizeContentUri(script.uri), _drawflowVariable, script.integrity, cancellationToken: token);
    }

    private async ValueTask InitializeInterop(CancellationToken token)
    {
        _ = await _moduleImportUtil.GetContentModuleReference(_modulePath, token);
    }

    private async ValueTask<IJSObjectReference> GetModule(CancellationToken cancellationToken)
    {
        return await _moduleImportUtil.GetContentModuleReference(_modulePath, cancellationToken);
    }

    private async ValueTask InvokeVoidAsync(string identifier, CancellationToken cancellationToken, params object?[] args)
    {
        IJSObjectReference module = await GetModule(cancellationToken);
        await module.InvokeVoidAsync(identifier, cancellationToken, args);
    }

    private async ValueTask<T> InvokeAsync<T>(string identifier, CancellationToken cancellationToken, params object?[] args)
    {
        IJSObjectReference module = await GetModule(cancellationToken);
        return await module.InvokeAsync<T>(identifier, cancellationToken, args);
    }

    public async ValueTask Initialize(bool useCdn, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            await _styleInitializer.Init(useCdn, linked);
            await _scriptInitializer.Init(useCdn, linked);
            await _interopInitializer.Init(linked);
        }
    }

    public async ValueTask Create(string elementId, DrawflowOptions? options = null, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            bool useCdn = options?.UseCdn ?? true;

            await _styleInitializer.Init(useCdn, linked);
            await _scriptInitializer.Init(useCdn, linked);
            await _interopInitializer.Init(linked);

            string? json = null;

            if (options != null)
                json = JsonUtil.Serialize(options);

            await InvokeVoidAsync("create", linked, elementId, json);
        }
    }

    public ValueTask AddNode(string elementId, string name, int inputs, int outputs, int posX, int posY, string className, object? data, string html, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeVoidAsync("addNode", linked, elementId, name, inputs, outputs, posX, posY, className, data, html);
        }
    }

    public async ValueTask AddNode(string elementId, DrawflowNode node, CancellationToken cancellationToken = default)
    {
        if (node.Name.IsNullOrEmpty())
            throw new ArgumentException("Node name cannot be null or empty", nameof(node));

        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            int inputs = node.Inputs?.Count ?? 0;
            int outputs = node.Outputs?.Count ?? 0;

            await InvokeVoidAsync("addNode", linked, elementId, node.Name, inputs, outputs, node.PosX, node.PosY, node.Class ?? "", node.Data, node.Html ?? "");
        }
    }

    public ValueTask RemoveNode(string elementId, string nodeId, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeVoidAsync("removeNode", linked, elementId, nodeId);
        }
    }

    public ValueTask AddConnection(string elementId, string outputNode, string inputNode, string outputClass, string inputClass, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeVoidAsync("addConnection", linked, elementId, outputNode, inputNode, outputClass, inputClass);
        }
    }

    public ValueTask<string> ExportAsJson(string elementId, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeAsync<string>("exportAsJson", linked, elementId);
        }
    }

    public async ValueTask<DrawflowExport> Export(string elementId, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            string json = await InvokeAsync<string>("exportFlow", linked, elementId);
            return JsonUtil.Deserialize<DrawflowExport>(json) ?? new DrawflowExport();
        }
    }

    public ValueTask Import(string elementId, string json, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeVoidAsync("importFlow", linked, elementId, json);
        }
    }

    public async ValueTask Import(string elementId, DrawflowExport drawflowExport, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            string? json = JsonUtil.Serialize(drawflowExport);
            await InvokeVoidAsync("importFlow", linked, elementId, json);
        }
    }

    public ValueTask Destroy(string elementId, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeVoidAsync("destroy", linked, elementId);
        }
    }

    public ValueTask CreateObserver(string elementId, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeVoidAsync("createObserver", linked, elementId);
        }
    }

    public ValueTask AddEventListener(string elementId, string eventName, EventCallback<string> callback, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            var dotNet = DotNetObjectReference.Create(new CallbackInvoker(callback));
            return InvokeVoidAsync("addEventListener", linked, elementId, eventName, dotNet);
        }
    }

    public ValueTask ZoomIn(string elementId, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeVoidAsync("zoomIn", linked, elementId);
        }
    }

    public ValueTask ZoomOut(string elementId, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeVoidAsync("zoomOut", linked, elementId);
        }
    }

    public ValueTask AddModule(string elementId, string name, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeVoidAsync("addModule", linked, elementId, name);
        }
    }

    public async ValueTask AddModule(string elementId, string moduleName, DrawflowModule module, CancellationToken cancellationToken = default)
    {
        if (moduleName.IsNullOrEmpty())
            throw new ArgumentException("Module name cannot be null or empty", nameof(moduleName));

        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            await InvokeVoidAsync("addModule", linked, elementId, moduleName);

            if (module.Data != null)
            {
                foreach (DrawflowNode node in module.Data.Values)
                    await AddNode(elementId, node, linked);
            }
        }
    }

    public ValueTask ChangeModule(string elementId, string name, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeVoidAsync("changeModule", linked, elementId, name);
        }
    }

    public ValueTask RemoveModule(string elementId, string name, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeVoidAsync("removeModule", linked, elementId, name);
        }
    }

    public async ValueTask<DrawflowNode?> GetNodeFromId(string elementId, string id, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            string json = await InvokeAsync<string>("getNodeFromId", linked, elementId, id);
            if (string.IsNullOrWhiteSpace(json))
                return null;

            return JsonUtil.Deserialize<DrawflowNode>(json);
        }
    }

    public ValueTask<List<string>> GetNodesFromName(string elementId, string name, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeAsync<List<string>>("getNodesFromName", linked, elementId, name);
        }
    }

    public ValueTask UpdateNodeData(string elementId, string id, object data, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeVoidAsync("updateNodeData", linked, elementId, id, data);
        }
    }

    public ValueTask AddNodeInput(string elementId, string id, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeVoidAsync("addNodeInput", linked, elementId, id);
        }
    }

    public ValueTask AddNodeOutput(string elementId, string id, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeVoidAsync("addNodeOutput", linked, elementId, id);
        }
    }

    public ValueTask RemoveNodeInput(string elementId, string id, string inputClass, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeVoidAsync("removeNodeInput", linked, elementId, id, inputClass);
        }
    }

    public ValueTask RemoveNodeOutput(string elementId, string id, string outputClass, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeVoidAsync("removeNodeOutput", linked, elementId, id, outputClass);
        }
    }

    public ValueTask RemoveSingleConnection(string elementId, string outId, string inId, string outClass, string inClass, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeVoidAsync("removeSingleConnection", linked, elementId, outId, inId, outClass, inClass);
        }
    }

    public ValueTask UpdateConnectionNodes(string elementId, string id, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeVoidAsync("updateConnectionNodes", linked, elementId, id);
        }
    }

    public ValueTask RemoveConnectionNodeId(string elementId, string id, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeVoidAsync("removeConnectionNodeId", linked, elementId, id);
        }
    }

    public ValueTask<string?> GetModuleFromNodeId(string elementId, string id, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeAsync<string?>("getModuleFromNodeId", linked, elementId, id);
        }
    }

    public ValueTask ClearModuleSelected(string elementId, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeVoidAsync("clearModuleSelected", linked, elementId);
        }
    }

    public ValueTask Clear(string elementId, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeVoidAsync("clear", linked, elementId);
        }
    }

    public ValueTask SetZoom(string elementId, double zoom, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeVoidAsync("setZoom", linked, elementId, zoom);
        }
    }

    public ValueTask<double> GetZoom(string elementId, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeAsync<double>("getZoom", linked, elementId);
        }
    }

    public ValueTask CenterNode(string elementId, string id, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeVoidAsync("centerNode", linked, elementId, id);
        }
    }

    public ValueTask<object> GetNodePosition(string elementId, string id, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeAsync<object>("getNodePosition", linked, elementId, id);
        }
    }

    public ValueTask SetNodePosition(string elementId, string id, int posX, int posY, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeVoidAsync("setNodePosition", linked, elementId, id, posX, posY);
        }
    }

    public ValueTask<string> GetNodeHtml(string elementId, string id, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeAsync<string>("getNodeHtml", linked, elementId, id);
        }
    }

    public ValueTask SetNodeHtml(string elementId, string id, string html, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeVoidAsync("setNodeHtml", linked, elementId, id, html);
        }
    }

    public ValueTask<string> GetNodeClass(string elementId, string id, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeAsync<string>("getNodeClass", linked, elementId, id);
        }
    }

    public ValueTask SetNodeClass(string elementId, string id, string className, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeVoidAsync("setNodeClass", linked, elementId, id, className);
        }
    }

    public ValueTask<string> GetNodeName(string elementId, string id, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeAsync<string>("getNodeName", linked, elementId, id);
        }
    }

    public ValueTask SetNodeName(string elementId, string id, string name, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeVoidAsync("setNodeName", linked, elementId, id, name);
        }
    }

    public ValueTask<List<object>> GetNodeConnections(string elementId, string id, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeAsync<List<object>>("getNodeConnections", linked, elementId, id);
        }
    }

    public ValueTask<bool> IsNodeSelected(string elementId, string id, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeAsync<bool>("isNodeSelected", linked, elementId, id);
        }
    }

    public ValueTask SelectNode(string elementId, string id, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeVoidAsync("selectNode", linked, elementId, id);
        }
    }

    public ValueTask UnselectNode(string elementId, string id, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeVoidAsync("unselectNode", linked, elementId, id);
        }
    }

    public ValueTask<List<string>> GetSelectedNodes(string elementId, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeAsync<List<string>>("getSelectedNodes", linked, elementId);
        }
    }

    public ValueTask ClearSelectedNodes(string elementId, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeVoidAsync("clearSelectedNodes", linked, elementId);
        }
    }

    public ValueTask<string> GetCurrentModule(string elementId, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeAsync<string>("getCurrentModule", linked, elementId);
        }
    }

    public ValueTask<List<string>> GetModules(string elementId, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeAsync<List<string>>("getModules", linked, elementId);
        }
    }

    public ValueTask<bool> IsEditMode(string elementId, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeAsync<bool>("isEditMode", linked, elementId);
        }
    }

    public ValueTask SetEditMode(string elementId, bool editMode, CancellationToken cancellationToken = default)
    {
        CancellationToken linked = _cancellationScope.CancellationToken.Link(cancellationToken, out CancellationTokenSource? source);

        using (source)
        {
            return InvokeVoidAsync("setEditMode", linked, elementId, editMode);
        }
    }

    public async ValueTask DisposeAsync()
    {
        await _moduleImportUtil.DisposeContentModule(_modulePath);

        await _interopInitializer.DisposeAsync();
        await _styleInitializer.DisposeAsync();
        await _scriptInitializer.DisposeAsync();

        await _cancellationScope.DisposeAsync();
    }

    private sealed class CallbackInvoker
    {
        private readonly EventCallback<string> _callback;
        public CallbackInvoker(EventCallback<string> callback) { _callback = callback; }
        [JSInvokable]
        public Task Invoke(string json) => _callback.InvokeAsync(json);
    }
}
